nemo/collections/asr/models/asr_model.py

```diff
@@ -136,9 +136,56 @@ def on_after_backward(self):
                 torch.distributed.all_reduce(valid_gradients, op=torch.distributed.ReduceOp.MIN)
 
             if valid_gradients < 1:
-                logging.warning(f'detected inf or nan values in gradients! Setting gradients to zero.')
+                logging.warning('detected inf or nan values in gradients! Setting gradients to zero.')
                 self.zero_grad()
 
+    def on_train_start(self):
+        super().on_train_start()
+
+        # dynamic freezing
+        # should fire only once, on the very first batch of training and never again
+        if not hasattr(self, '_freeze_cfg'):
+            if (
+                hasattr(self.cfg, 'freeze_updates')
+                and self.cfg.freeze_updates is not None
+                and self.cfg.freeze_updates.get('enabled', False)
+            ):
+                setattr(self, '_freeze_cfg', OmegaConf.to_container(self.cfg.freeze_updates))
+                self._freeze_cfg['is_frozen'] = {k: False for k in self._freeze_cfg['modules'].keys()}
+            else:
+                setattr(self, '_freeze_cfg', None)
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int):
+        """
+        dynamic freezing
+        allows you to freeze certain model layers for the first N steps of training
+        """
+        super().on_train_batch_start(batch, batch_idx)
+
+        # dynamic freezing
+        if hasattr(self, '_freeze_cfg') and self._freeze_cfg is not None:
+            if self.training and hasattr(self, "trainer") and self.trainer is not None:
+                num_updates = self.trainer.global_step + 1
+
+                for ml, m_steps in self._freeze_cfg['modules'].items():
+                    # we could do hasattr check here, but it's too expensive for each step
+                    # consequently you'll throw an error if the module name doesn't exist
+                    # or was spelled wrong in the config.yaml
+                    if num_updates <= m_steps and not self._freeze_cfg['is_frozen'][ml]:
```

Comment:
Other suggestion which can be done later in future is to add a support of passing start and end steps for freezing like this. -1 means to the end:
  freeze_updates:
    enabled: true

```
    modules:
      preprocessor: 50
      encoder: [10, 500]
      joint: [600, -1]

```

Reply:
I agree this would be something good to add later down the line. I'm curious to first release this initial version and see if people use it and find it helpful, and then maybe add this if there is demand for it? Or would you prefer I add this functionality from the getgo?

Reply:
I'd prefer from the get go. It would be a breaking change later

Reply:
Is this something we really need though? Have there been any papers or methodologies ever published wherein people would want to freeze layers in such a schedule form? Like the [10, 500] case, I can't really wrap my head around why that would ever be useful.

Reply:
There may be cases where you may want to freeze them after some steps, then you may need to use something like [50,-1]. 
We can avoid later breaking change by supporting both cases. When it is a simple integer, it specifies the start and if it is a list it would specify a span?

Reply:
Okay, I will implement this, but I don't understand what the [50, -1] logic is supposed to do. Does that mean you freeze for the first 50 steps, or you freeze after you reach 50 and until the end (so only the first 50 steps are UNfrozen)?

Reply:
[50,-1] means start freezing at step 50 until the end. If there is a single integer, it means the start of freezing but when it is a list, it specifies a range where it is freezed.

Reply:
I haved committed new code which does this exactly as you have asked for